stm32: Added OSPI support for STM32H7 series.

nspsck opened the pull request:
Added OSPI support for STM32H7 series. As well as
fixed some misslabled pins for STM32H7B3I_DK.

This also closes #12517 .

nspsck committed:
stm32/boards/STM32H7B3I_DK: Fix octo-spi pin configuration.

The original OSPIFLASH settings in the `mpconfigboard.h` contained some
mistakes that prevented the firmware from compiling.  These are now
corrected and the firmware can be built with support for OSPI flash.

Note: external storage in OSPI flash is not yet configured on this board.

Signed-off-by: nspsck <teng.jiang94@gmail.com>

nspsck committed:
stm32/octospi: Add OSPI support for STM32H7 MCUs.

Added a if-statement to `octospi.c` to detect if the targeted MCU is one of
the STM32H7 series.  If that is the case, another set of variables are used
for the `mp_hal_pin_config_alt_static_speed()` function, as well as for
register `OCTOSPI1->CR`.  This allows the STM32H723 and STM32H7B3 series
MCU to use octo-spi flash like the STM32H573 series MCU.

Signed-off-by: nspsck <teng.jiang94@gmail.com>

JohnieBraaf commented:
I wanted to test this, but somehow whatever firmware version I use I'm not getting REPL anymore.
But good to see this finalized anyway.

nspsck commented:
Hi JohnieBraaf

Which board are you using? Are you sure that the PLLs are set correctly? Typically, if you were trying to access REPL from the USB port, you have to set the USB freqquency to 48MHz, otherwise you can not establish a connection to your device. (Just FYI, in case you missed this)

These changes should not conflict with anything REPL related. Since only some variable's names or some pin's names has been changed. But it's only a `should` :P

JohnieBraaf commented:
@nspsck, thnx for the tip, I'm using the STM32H7B3I_DK, pretty sure it was working before..., I might have to try STMCube to check clocks and see if the board might have died on me.

nspsck commented:
@JohnieBraaf That's weird, well, maybe you can set that one line (line 42) in the file:

`ports/stm32/boards/STM32H7B3I_DK/mpconfigboard.h`

from `if 1` to `if 0`. Maybe by some magic that would help.

JohnieBraaf commented:
It was my bad, I forgot I was using ST-link for REPL as stated in (https://github.com/micropython/micropython/commit/9e2423e730ae32cd82e5c06099e6a847a2020f6a). REPL is working as expected now.

Thnx for the pointers you gave on Discord. I will be trying to get the OSPI flash configured on the STM32H73BI_DK board now.

JohnieBraaf commented:
Today I was able to test this PR and I can confirm it's working for my STM32H7B3I_DK board.
Not sure what is wise to do, submit my board configuration in your branch or make a separate PR once this one is merged.

Anyways, this was my test:

```
MicroPython 1c3caf596-dirty on 2023-10-18; STM32H7B3I-DK with STM32H7B3LIH6Q
Type "help()" for more information.
>>> import os
>>> os.getcwd()
'/flash'
>>> os.mkdir('test')
>>> os.listdir('/flash')
['boot.py', 'main.py', 'pybcdc.inf', 'README.txt', 'test']
>>> with open('/flash/hello.txt', 'w') as f:
...     f.write('Hello world')
...
...
11
>>> print(open('/flash/hello.txt').read())
Hello world
```

Everything seems to work as intended. Many thanks!

nspsck commented:
That's great!

On wether you should create a separate branch for the config file:

You should probably discuss it with one of the maintainers. 

Me personally would have waited for this to merged, cause I am a "one thing after another" kind of guy.

JohnieBraaf commented:
That indeed sounds like the best way forward. I will create a separate PR and mention that it is dependent on this one.

dpgeorge commented on ports/stm32/boards/STM32H7B3I_DK/mpconfigboard.h:
```diff
@@ -39,20 +39,20 @@
+#define MICROPY_HW_OSPIFLASH_SCK    (pin_B2)
 #define MICROPY_HW_OSPIFLASH_DQS    (pin_C5)
-#define MICROPY_HW_OSPIFLASH_IO0    (pin_P8)
+#define MICROPY_HW_OSPIFLASH_IO0    (pin_D11) // changed from pin_P8
```
No need to have this comment about it being changed, just change it.

dpgeorge commented on ports/stm32/boards/STM32H7B3I_DK/mpconfigboard.h:
```diff
@@ -39,20 +39,20 @@
 #define MICROPY_HW_PWR_SMPS_CONFIG  (PWR_DIRECT_SMPS_SUPPLY)
 
-#if 0
+#if 1
```
This could be `!MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE` so that one can select between internal or external flash storage.

dpgeorge commented on ports/stm32/octospi.c:
```diff
@@ -45,6 +45,24 @@
 void octospi_init(void) {
     // Configure OCTOSPI pins (allows 1, 2, 4 or 8 line configuration).
+    #if defined(STM32H7)
+    mp_hal_pin_config_alt_static_speed(MICROPY_HW_OSPIFLASH_CS, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_OCTOSPIM_P1_NCS);
```
Duplicating the code like this is no good.  Instead, I suggest making macros for the STATIC_AF part.  Eg:
```
#if defined(STM32H7)
#define STATIC_AF_OCTOSPI(signal) STATIC_AF_OCTOSPIM_P1_ ## signal
#else
#define STATIC_AF_OCTOSPI(signal) STATIC_AF_OCTOSPI1_ ## signal
#endif
```
then use it like `STATIC_AF_OCTOSPI(CLK)`.

nspsck replied:
Thank you very much for your patience going through the codes and all the elegant suggestions to it! 

All of these changes has been done by now and I have made some changes to `system_stm32.c` to allow the clocksource selection. 

Tho, even I've tried to follow the instructions here [https://github.com/jimmo/git-and-micropython#my-pr-has-multiple-commits-2](https://github.com/jimmo/git-and-micropython#my-pr-has-multiple-commits-2) as much as I could, I think I've messed up the commit messages... If that was the case, please tell me what I need to do!

nspsck committed:
stm32/system_stm32: Allow selection of OSPI clock source.

Added a #if-block to `system_stm32.c` to check whether
`MICROPY_HW_RCC_OSPI_CLKSOURCE` is defined.  If that is the case, the
clock source for the OSPI will be changed to the specified source.

Signed-off-by: nspsck <teng.jiang94@gmail.com>

roizano commented:
Hi ,

Found the next issue in the core code of STM32 boards while compiling STM32H750 :

from: https://github.com/jahr/STM32H750B_DK?tab=readme-ov-file#stm32h750b_dk 

```
CC boards/WEACT_H750/bdev.c
boards/WEACT_H750/bdev.c:4:7: error: expected ';' before 'const'
    4 | STATIC const mp_soft_spi_obj_t soft_spi_bus = {
      |       ^~~~~~
      |       ;
boards/WEACT_H750/bdev.c:13:1: error: unknown type name 'STATIC'
   13 | STATIC mp_spiflash_cache_t spi_bdev_cache;
      | ^~~~~~
boards/WEACT_H750/bdev.c:13:28: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'spi_bdev_cache'
   13 | STATIC mp_spiflash_cache_t spi_bdev_cache;
      |                            ^~~~~~~~~~~~~~
boards/WEACT_H750/bdev.c:21:15: error: 'spi_bdev_cache' undeclared here (not in a function); did you mean 'spi_bdev_ioctl'?
   21 |     .cache = &spi_bdev_cache,
      |               ^~~~~~~~~~~~~~
      |               spi_bdev_ioctl
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
../../py/mkrules.mk:90: recipe for target 'build-WEACT_H750/boards/WEACT_H750/bdev.o' failed
make: *** [build-WEACT_H750/boards/WEACT_H750/bdev.o] Error 1
```
someone knows what could cause this issue?

nspsck commented:
You should change all `STATIC` to `static`. Or use an older version of mpy. This will solve your problem.

The reason was that mpy changed their 'STATIC' macro on demand of some sponsors (if I remember correctly) a while ago.

dpgeorge commented:
@nspsck thanks for updating this PR and sorry it took some time to come back to.

I've now rebased it on latest master and merged.

@JohnieBraaf feel free to open a PR to enable external OSPI flash storage on the `STM32H7B3I_DK` board.
